File: src/api/scheduled-article.js
import Vue from "vue";

import getProp from "@/utils/getter.js";
import imgproxyURL from "@/api/imgproxy-url.js";

export default class ScheduledArticle {
  constructor({ id, data, client }) {
    this._url_id = id;
    this._client = client;

    this.init(data);

    Vue.observable(this);
  }

  init(data, { save_reset = true } = {}) {
    this.isSaving = false;
    this.saveError = null;
    this._refreshArc = false;

    if (save_reset) {
      this._reset = JSON.stringify(data);
    }

    let props = {
      id: ["InternalID", ""],
      arcID: ["ArcID", ""],
      body: ["Body", ""],
      blurb: ["Blurb", ""],
      byline: ["Byline", ""],
      hed: ["Hed", ""],
      imageDescription: ["ImageDescription", ""],
      imageCaption: ["ImageCaption", ""],
      imageCredit: ["ImageCredit", ""],
      imageURL: ["ImageURL", ""],
      imageSize: ["ImageSize", ""],
      linkTitle: ["LinkTitle", ""],
      slug: ["Slug", ""],
      subhead: ["Subhead", ""],
      summary: ["Summary", ""],
      kicker: ["Kicker", ""],
      suppressFeatured: ["SuppressFeatured", false],
      weight: ["Weight", 0],
      overrideURL: ["OverrideURL", ""],
      aliases: ["Aliases", []],
      modalExclude: ["ModalExclude", false],
      noIndex: ["NoIndex", false],

      warnings: ["Warnings", []],
      authors: ["Authors", []],
      series: ["Series", []],
      topics: ["Topics", []],
    };

    for (let [key, [val, fallback]] of Object.entries(props)) {
      this[key] = getProp(data, val, { fallback });
    }

    // Date getters
    props = {
      scheduleFor: "ScheduleFor",
      lastArcSync: "LastArcSync",
      pubDate: "PubDate",
      lastSaved: "LastSaved",
      lastPublished: "LastPublished",
    };

    let dateObj = {};

    for (let [key, val] of Object.entries(props)) {
      dateObj[key] = getProp(data, val, { fallback: null });
    }

    for (let [prop] of Object.entries(props)) {
      let val = dateObj[prop];
      this[prop] = val ? new Date(val) : null;
    }
  }

  reset() {
    this.init(JSON.parse(this._reset), { save_reset: false });
  }

  toString() {
    return `Scheduled Article ${this.id}`;
  }

  toJSON() {
    return {
      InternalID: this.id,
      ArcID: this.arcID,
      Body: this.body,
      Byline: this.byline,
      Blurb: this.blurb,
      Hed: this.hed,
      ImageCaption: this.imageCaption,
      ImageCredit: this.imageCredit,
      ImageDescription: this.imageDescription,
      ImageURL: this.imageURL,
      LinkTitle: this.linkTitle,
      Slug: this.slug,
      Subhead: this.subhead,
      Summary: this.summary,
      Authors: this.authors,
      Topics: this.topics,
      Series: this.series,
      ScheduleFor: this.scheduleFor,
      LastArcSync: this.lastArcSync,
      PubDate: this.pubDate,
      Kicker: this.kicker,
      SuppressFeatured: this.suppressFeatured,
      ImageSize: this.imageSize,
      OverrideURL: this.overrideURL,
      Aliases: this.aliases,
      ModalExclude: this.modalExclude,
      NoIndex: this.noIndex,
      "almanack-refresh-arc": this._refreshArc,
    };
  }

  deriveSlug() {
    this.slug = this.hed
      .toLowerCase()
      .replace(/\b(the|an?)\b/g, " ")
      .replace(/\bpa\b/g, "pennsylvania")
      .replace(/\W+/g, " ")
      .trim()
      .replace(/ /g, "-");
  }

  get hasPublished() {
    return !!this.lastPublished;
  }

  get pubURL() {
    if (!this.slug) {
      return "";
    }
    let year = this.pubDate.getFullYear();
    let month = (this.pubDate.getMonth() + 1).toString().padStart(2, "0");
    return `https://www.spotlightpa.org/news/${year}/${month}/${this.slug}/`;
  }

  get imagePreviewURL() {
    if (!this.imageURL || this.imageURL.match(/^http/)) {
      return "";
    }
    return imgproxyURL(this.imageURL);
  }

  validate() {
    let valid = true;
    if (!this.kicker) {
      this.saveError = new Error("Kicker must not be blank");
      this.saveError.name = "Validation Error";
      valid = false;
    }
    if (!this.imageURL) {
      this.saveError = new Error("Featured image must not be blank");
      this.saveError.name = "Featured image error";
      valid = false;
    }
    if (!this.slug) {
      this.saveError = new Error("Article slug must not be blank");
      this.saveError.name = "Validation Error";
      valid = false;
    }
    return valid;
  }

  async save({ schedule = null, refreshArc = false }) {
    if (schedule) {
      this.scheduleFor = this.pubDate;
    } else if (schedule !== null) {
      this.scheduleFor = null;
    }
    if (this.scheduleFor && !this.validate() && !refreshArc) {
      return;
    }
    this.isSaving = true;
    this._refreshArc = refreshArc;

    let data;
    [data, this.saveError] = await this._client.saveArticle(this);
    this.isSaving = false;
    this._refreshArc = false;
    if (!this.saveError) {
      this.init(data);
      this.validate();
    }
  }
}

File: pkg/almanack/scheduled-article.go
package almanack

import (
	"encoding/json"
	"time"

	"github.com/spotlightpa/almanack/internal/db"
)

type ScheduledArticle struct {
	Article
	LastArcSync   time.Time
	ScheduleFor   *time.Time
	LastSaved     *time.Time
	LastPublished *time.Time
}

func (schArticle *ScheduledArticle) toSPLData() interface{} {
	type splDataType struct {
		ID               *string     `json:"internal-id"`
		Budget           *string     `json:"budget"`
		ImageURL         *string     `json:"image-url"`
		ImageCaption     *string     `json:"image-caption"`
		ImageCredit      *string     `json:"image-credit"`
		PubDate          *time.Time  `json:"pub-date"`
		Slug             *string     `json:"slug"`
		Authors          *[]string   `json:"authors"`
		Byline           *string     `json:"byline"`
		Hed              *string     `json:"hed"`
		Subhead          *string     `json:"subhead"`
		Summary          *string     `json:"summary"`
		Blurb            *string     `json:"blurb"`
		Kicker           *string     `json:"kicker"`
		Body             *string     `json:"body"`
		LinkTitle        *string     `json:"link-title"`
		SuppressFeatured *bool       `json:"suppress-featured"`
		LastArcSync      *time.Time  `json:"last-arc-sync"`
		LastSaved        **time.Time `json:"last-saved"`
	}
	return &splDataType{
		ID:               &schArticle.InternalID,
		Budget:           &schArticle.Budget,
		ImageURL:         &schArticle.ImageURL,
		ImageCaption:     &schArticle.ImageCaption,
		ImageCredit:      &schArticle.ImageCredit,
		PubDate:          &schArticle.PubDate,
		Slug:             &schArticle.Slug,
		Authors:          &schArticle.Authors,
		Byline:           &schArticle.Byline,
		Hed:              &schArticle.Hed,
		Subhead:          &schArticle.Subhead,
		Summary:          &schArticle.Summary,
		Blurb:            &schArticle.Blurb,
		Kicker:           &schArticle.Kicker,
		Body:             &schArticle.Body,
		LinkTitle:        &schArticle.LinkTitle,
		SuppressFeatured: &schArticle.SuppressFeatured,
		LastArcSync:      &schArticle.LastArcSync,
		LastSaved:        &schArticle.LastSaved,
	}
}

func (schArticle *ScheduledArticle) fromDB(dbArticle db.Article) error {
	schArticle.ArcID = dbArticle.ArcID.String
	schArticle.ScheduleFor = timeNull(dbArticle.ScheduleFor)
	schArticle.LastPublished = timeNull(dbArticle.LastPublished)
	schArticle.filepath = dbArticle.SpotlightPAPath.String

	if err := json.Unmarshal(dbArticle.SpotlightPAData, schArticle.toSPLData()); err != nil {
		return err
	}

	if schArticle.LastArcSync.IsZero() {
		if err := schArticle.ResetArcData(dbArticle); err != nil {
			return err
		}
	}
	return nil
}

func (schArticle *ScheduledArticle) ResetArcData(dbArticle db.Article) error {
	schArticle.LastArcSync = time.Now()
	var arcStory ArcStory
	if err := json.Unmarshal(dbArticle.ArcData, &arcStory); err != nil {
		return err
	}
	art, err := arcStory.ToArticle()
	if err != nil {
		return err
	}
	schArticle.Article = *art
	return nil
}

func (schArticle *ScheduledArticle) toDB() (*db.Article, error) {
	var dart db.Article

	dart.ArcID = nullString(schArticle.ArcID)
	dart.ScheduleFor = nullTime(schArticle.ScheduleFor)
	dart.LastPublished = nullTime(schArticle.LastPublished)
	dart.SpotlightPAPath = nullString(schArticle.ContentFilepath())

	var err error
	if dart.SpotlightPAData, err = json.Marshal(schArticle.toSPLData()); err != nil {
		return nil, err
	}
	return &dart, nil
}
